docs/zkapps/o1js/bitwise-gadgets.mdx

```diff
@@ -0,0 +1,320 @@
+- The and gadget works by comparing two bits and returns `1` only if both bits are `1`, otherwise it returns `0`.
+- Constrains both input values to fit into `paddedLength` bits. 
+- Both Field elements must fit into `2^paddedLength - 1` or an error is thrown.
+- Guarantees that the output is has at most `paddedLength` bits.
+- Can be checked by a double generic gate that verifies the following relationship between the values. In the process, it also invokes the [xor()](#xor) gadget that creates additional constraints depending on `length`.
+
+  The generic gate verifies `a + b = sum` and the conjunction equation `2 * and = sum - xor`.
+
+  where:
+  * `a + b = sum`
+  * `a ^ b = xor`
+  * `a & b = and`
+
+For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.
+
+The `length` parameter:
+
+- Defines how many bits to compare. 
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- Specifying a larger `length` parameter adds additional constraints.
+
+With `length = 2` (`paddedLength = 16`), the `and()` operation fails for any input that is larger than `2**16`.
+
+Example:
+
+```ts
+let a = Field(3);    // ... 000011
+let b = Field(5);    // ... 000101
+let c = Gadgets.and(a, b, 2);    // ... 000001
+c.assertEquals(1);
+```
+
+## compactMultiRangeCheck
+
+```ts
+let [x, y] = Gadgets.compactMultiRangeCheck([xy, z]);
+```
+
+The bitwise compactMultiRangeCheck gadget is a variant of [multiRangeCheck](#multirangecheck) where the first two variables are passed in combined form `xy = x + 2^88*y`. It is similar to the [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operation in JavaScript. ,
+
+The compactMultiRangeCheck gadget:
+
+- Splits `xy` into `x` and `y`.
+- Proves that `xy = x + 2^88*y`.
+- Proves that `x, y, z` are all in the range `[0, 2^88)`.
+- Returns the split form `[x, y, z]`.
+- Throws an error if `xy` exceeds `2*88 = 176` bits or if `z` exceeds 88 bits.
+
+Example:
+
+```ts
+compactMultiRangeCheck(xy: Field, z: Field) {
+return compactMultiRangeCheck(xy, z);
+},
+```
+
+## leftShift
+
+```ts
+leftShift(field: Field, bits: number) {
+    return leftShift(field, bits);
+},
+```
+
+The leftShift gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the left. Unlike rotation, the bits that fall off at the end are discarded and the vacant positions are filled with zeros.
+
+The leftShift gadget:
+
+- Performs a left shift operation on a Field element.
+- Shifts bits to the left and discards the overflowing bits.
+- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
+- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).
```

Comment:
can use or must use?

Reply:
"Can use" is fine here and in the other instances.

Using rangecheck before leftshift etc is not a prerequisite.